arangod/Replication2/StateMachines/Document/DocumentLeaderState.cpp

```diff
@@ -98,54 +94,79 @@ auto DocumentLeaderState::recoverEntries(std::unique_ptr<EntryIterator> ptr)
+                      << "will not apply transaction " << op.tid
+                      << " for shard " << op.shard
+                      << " because it is not available";
+                  return {};
+                } else if (res.fail()) {
+                  return res;
+                }
+                activeTransactions.insert(op.tid);
+                return transactionHandler->applyEntry(op);
+              },
+              [&](FinishesUserTransactionOrIntermediate auto& op) -> Result {
+                // There are two cases where we can end up here:
+                // 1. After recovery, we did not get the beginning of the
+                // transaction.
+                // 2. We ignored all other operations for this transaction
+                // because the shard was dropped.
+                if (activeTransactions.erase(op.tid) == 0) {
+                  return Result{};
+                }
+                return transactionHandler->applyEntry(op);
+              },
+              [&](ReplicatedOperation::DropShard& op) -> Result {
+                // Abort all transactions for this shard.
+                for (auto const& tid :
+                     transactionHandler->getTransactionsForShard(op.shard)) {
+                  activeTransactions.erase(tid);
+                  auto abortRes = transactionHandler->applyEntry(
+                      ReplicatedOperation::buildAbortOperation(tid));
+                  if (abortRes.fail()) {
+                    LOG_CTX("3eb75", INFO, self->loggerContext)
+                        << "failed to abort transaction " << tid
+                        << " for shard " << op.shard
+                        << " during recovery: " << abortRes;
+                    return abortRes;
+                  }
+                }
+                return transactionHandler->applyEntry(op);
+              },
+              [&](auto&& op) { return transactionHandler->applyEntry(op); }},
+          doc.getInnerOperation());
+
+      if (res.fail()) {
+        LOG_CTX("cbc5b", FATAL, self->loggerContext)
+            << "failed to apply entry " << doc << " during recovery: " << res;
+        FATAL_ERROR_EXIT();
       }
     }
 
-    auto doc = DocumentLogEntry{{},  // this affects all shards
-                                OperationType::kAbortAllOngoingTrx,
-                                {},
-                                TransactionId{0},
-                                {}};
-    auto stream = self->getStream();
-    stream->insert(doc);
+    auto abortAll = ReplicatedOperation::buildAbortAllOngoingTrxOperation();
+    auto abortAllRes =
+        self->replicateOperation(abortAll, ReplicationOptions{}).get();
+    if (abortAllRes.fail()) {
+      LOG_CTX("b4217", FATAL, self->loggerContext)
+          << "failed to replicate AbortAllOngoingTrx operation during "
+             "recovery: "
+          << abortAllRes.result();
+      FATAL_ERROR_EXIT();
```

Comment:
Do we really need to block here until the entry is replicated?

Reply:
I don't think so.

Reply:
We are not actually blocking, because we don't wait for commit. I will add an assertion to make sure that the future is always ready after the `replicateOperation` call.

Reply:
ah because `waitForCommit` is not set in the options 😉